controllers/suite_test.go

```diff
@@ -44,6 +47,12 @@ var k8sClient client.Client
 var testEnv *envtest.Environment
 var reconciler *ProgressiveRolloutReconciler
 
+type mockArgoCDAppClient struct{}
+
+func (*mockArgoCDAppClient) Sync(ctx context.Context, in *applicationpkg.ApplicationSyncRequest, opts ...grpc.CallOption) (*argov1alpha1.Application, error) {
```

Comment:
I'm not sure I understand we we need to redefine the `mockArgoCDAppClient` and the sync method here. Can't we import it from a package?

Reply:
From what I saw, defining mocks in the test files they're used seems to be the most common practice (see https://github.com/kubernetes/kubernetes/blob/cea1d4e20b4a7886d8ff65f34c6d4f95efcb4742/staging/src/k8s.io/apiserver/pkg/authentication/token/union/unionauth_test.go#L30, for an example). 

But if we prefer to have a package for the mocks, I'm happy to move it.

Reply:
I don't think I have good enough knowledge on the ecosystem and golang on this to be opinionated :)

I guess my point it was to be able to reuse the mock instead of redefining it multiple times? I guess when I see `mockArgoCDAppClientCounter`, `mockArgoCDAppClient` and `mockArgoCDAppClientSyncOK` they _feel_ the same thing to me with different methods? But maybe not.

@sledigabel what do you think?

Reply:
They serve different purposes in different tests, but are all implementing the same interface:

1. `mockArgoCDAppClient` is not a *mock*, it really is a generic *stub* so that we can run the other tests when we don't care about what happens in the calls to Argo CD.
2. `mockArgoCDAppClientCounter` is really just a way to `assertCalledWith` the `Sync()` method - we could use this mock as the 'general' one (in 1.) if we so wished, I opted to make the other entirely useless to make it clear we _don't_ care about its functionality.
3. `mockArgoCDAppClientSyncOK` is used in unit testing to mock that the Argo CD app client returned OK, and conversely, `SyncNotOK` returns an error. 

There's good things with this approach, imo: it's very clear what each mock does, for instance. There's also bad things: if we ever opt to use one more method from the Argo CD App Client in the code, we have to update all these mocks, which adds toil to future maintainers.

Thoughts from those with more knowldege of Golang are most welcome here 🙂

Reply:
why bother creating your own mocks in the first place? Go can generate mocks for you (https://github.com/golang/mock) based on your interfaces.

Reply:
Don't see any reason not to use it, but I've never used any way of mocking in Golang, so I'm not opinionated on it.

Taking a step back then: what is the approach we want to take w.r.t. mocking in this project? If it's using `gomock`, then let's change these mocks :). Again, I'm not opinionated either way, but would rather decide on one approach and have consistency going forward.

Reply:
yeah i was going to suggest having a /mock package would be convenient, in case we need to reuse this for other tests moving forward.
Not sure about using mock to do that, but why not if that helps in the future.

Reply:
Alright - so I did some playground stuff with gomock and I think that might be useful in the long term. Bearing in mind that the Sync is a blocker to other tasks, I've opted to simply move the mocks to a `mocks` package for the time being. We should still agree on how we're doing these things long-term.